src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs

```diff
@@ -24,25 +30,48 @@ private static unsafe string GenerateId(long id)
             // and ~600% faster than calling long.ToString() on x86 in tight loops of 1 million+ iterations
             // See: https://github.com/aspnet/Hosting/pull/385
 
+#if NETCOREAPP2_1
+            return string.Create(13, id, _spanAction);
+#else
             // stackalloc to allocate array on stack rather than heap
             char* charBuffer = stackalloc char[13];
+            ref char encodeChars = ref _encode32Chars[0];
 
-            charBuffer[0] = _encode32Chars[(int)(id >> 60) & 31];
-            charBuffer[1] = _encode32Chars[(int)(id >> 55) & 31];
-            charBuffer[2] = _encode32Chars[(int)(id >> 50) & 31];
-            charBuffer[3] = _encode32Chars[(int)(id >> 45) & 31];
-            charBuffer[4] = _encode32Chars[(int)(id >> 40) & 31];
-            charBuffer[5] = _encode32Chars[(int)(id >> 35) & 31];
-            charBuffer[6] = _encode32Chars[(int)(id >> 30) & 31];
-            charBuffer[7] = _encode32Chars[(int)(id >> 25) & 31];
-            charBuffer[8] = _encode32Chars[(int)(id >> 20) & 31];
-            charBuffer[9] = _encode32Chars[(int)(id >> 15) & 31];
-            charBuffer[10] = _encode32Chars[(int)(id >> 10) & 31];
-            charBuffer[11] = _encode32Chars[(int)(id >> 5) & 31];
-            charBuffer[12] = _encode32Chars[(int)id & 31];
+            FillBuffer(ref Unsafe.AsRef<char>(charBuffer), id, ref encodeChars);
 
             // string ctor overload that takes char*
             return new string(charBuffer, 0, 13);
+#endif
+        }
+
+#if NETCOREAPP2_1
+        private static readonly SpanAction<char, long> _spanAction = FillBufferAction;
+
+        private static void FillBufferAction(Span<char> charBuffer, long id)
+        {
+            ref char buffer = ref MemoryMarshal.GetReference(charBuffer);
+            ref char encodeChars = ref _encode32Chars[0];
+
+            FillBuffer(ref buffer, id, ref encodeChars);
+        }
+#endif
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void FillBuffer(ref char buffer, long id, ref char encodeChars)
```

Comment:
Can't you get rid of the bounds check without using unsafe?

Reply:
Hm, I don't know how.

Even if the `char[]`-field (`_encode32Chars`) will be assigned to a local, the JIT has to emit a bound-check due the "random" access to this array.

For the `charBuffer` it isn't a problem.

Reply:
TBH, using unsafe code for this little of a gain isn't very attractive. We want to start removing unsafe code from Kestrel and using span in more places.

Reply:
> using span in more places.

Is definitely good. 👍 

But as long the JIT has to emit bound-checks for span there will be some unsafe parts. Span and unsafe is not xor :wink: I know what you mean...thx.